common/src/balance_unit.rs

```diff
@@ -0,0 +1,570 @@
+
+// Copyright (c) 2020, 2021, Polka Biome Ltd. All rights reserved.
+// SPDX-License-Identifier: BSD-4-Clause
+
+// Redistribution and use in source and binary forms, with or without modification,
+// are permitted provided that the following conditions are met:
+
+// Redistributions of source code must retain the above copyright notice, this list
+// of conditions and the following disclaimer.
+// Redistributions in binary form must reproduce the above copyright notice, this
+// list of conditions and the following disclaimer in the documentation and/or other
+// materials provided with the distribution.
+//
+// All advertising materials mentioning features or use of this software must display
+// the following acknowledgement: This product includes software developed by Polka Biome
+// Ltd., SORA, and Polkaswap.
+//
+// Neither the name of the Polka Biome Ltd. nor the names of its contributors may be used
+// to endorse or promote products derived from this software without specific prior written permission.
+
+// THIS SOFTWARE IS PROVIDED BY Polka Biome Ltd. AS IS AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL Polka Biome Ltd. BE LIABLE FOR ANY
+// DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+// BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use crate::prelude::FixedWrapper;
+use crate::{balance, Balance};
+use codec::{Decode, Encode, MaxEncodedLen};
+use core::cmp::Ordering;
+use core::ops::{Add, AddAssign, Div, Mul, Sub, SubAssign};
+use fixnum::ArithmeticError;
+use sp_arithmetic::traits::IntegerSquareRoot;
+use sp_runtime::traits::Zero;
+
+#[cfg(feature = "std")]
+use {
+    serde::{Deserialize, Serialize},
+    sp_std::fmt::Display,
+    static_assertions::_core::fmt::Formatter,
+};
+
+/// BalanceUnit wraps Balance and provides proper math operations between divisible & non-divisible balances that have different precision.
+#[derive(
+    Encode, Decode, Copy, Clone, Debug, PartialEq, Eq, scale_info::TypeInfo, MaxEncodedLen,
+)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+pub struct BalanceUnit {
+    inner: Balance,
+
+    /// Defines how to interpret the `inner` value.
+    /// true - means the precision of 18 digits
+    /// false - means the precision of 0 digits
+    is_divisible: bool,
+}
+
+impl Ord for BalanceUnit {
+    fn cmp(&self, other: &Self) -> Ordering {
+        let (left, right) = match (self.is_divisible, other.is_divisible) {
+            (false, true) => (balance!(self.inner), other.inner),
```

Comment:
- There is no way to get an integer part of a fixed.
- `Ord` trait doesn't provide an ability to return an error, we have only panic here.
- We already use `balance!()` in different parts of our project

Reply:
* I think it is doable, let me check
* If it's not possible then this trait should not be implemented
* it is used only in tests/benchmarks (with a few exceptions)

Reply:
Regarding retrieval of integer part:
* `other.inner / 1_000_000_000_000_000_000` (or it's better to use `FixedPrecision` somehow)
* `fixed.integral(Floor)`

Reply:
But it works wrong, because if we have, for example indivisible `3` and divisible `3.1111`, then after your suggestion the function returns `Ordering::Equal`, because we will have the case `3 == 3`

Reply:
> But it works wrong, because if we have, for example indivisible `3` and divisible `3.1111`, then after your suggestion the function returns `Ordering::Equal`, because we will have the case `3 == 3`

Yes, therefore I suggested to take the remainder into account. For example, one could compute
```rust
let quotient = other.inner / 1_000_000_000_000_000_000;
let remainder = other.inner % 1_000_000_000_000_000_000;
```